### PR TITLE
ci: introduce Release Please for automated release management

### PR DESCRIPTION
**Pull request by knqyf263**
## Motivation
Currently, releases are performed manually using `git tag` and `git push` in the local maintainer's environment. While this process has been functioning, it comes with certain risks and challenges:

1. Manual releases are prone to human error, leading to incorrect tags or other issues.
2. The timing of releases can be unclear, as it depends on manual action.
3. Difficulty in collaboration between maintainers for release.

To address these concerns and streamline our release process, I propose introducing [Release Please](https://github.com/googleapis/release-please) into our workflow.

## Proposed Changes
By integrating Release Please, we can automate key aspects of our release process:

1. Release Please will automatically create a release PR when a releasable commit (prefixed with `feat:` or `fix:`) is merged into a release branch (`main` and `release/v*`).
2. The release PR will contain the proposed version bump (minor for `feat:`, patch for `fix:`) and automatically generated release notes.
4. Once the release PR is approved and merged, Release Please will automatically create the version tag and perform the release.

This automation reduces the risk of human error and makes the release timing clear and predictable.

## Example
CleanShot 2024-05-27 at 23 03 19

**NOTE: I used my PAT so it looks like I created the PR, but in reality it is created by a bot.**

## Proposed Release Flow
The proposed release flow with Release Please will be as follows:

1. Merge releasable commits into the release branch (`main` and `release/v*`).
2. Release Please automatically creates a release PR with auto-generated release notes.
3. Draft the release notes in GitHub Discussions, as currently, according to the generated release notes in the release PR.
5. Approve and merge the release PR.
6. GitHub Actions automatically creates the version tag
7. GoReleaser performs the release.
8. Update the release notes in GitHub Discussions.

1, 2, 4 and 5 are new steps. Others are the same as the current flow.

For full details on the behavior of Release Flow, please see the Release Flow documentation.

## Benefits
- Reduced risk of human error in the release process.
- Easy to see what commits will be included in the next release.
- Release a new version if the maintainer approves it.
- Clear and predictable release timing.
- Streamlined and automated release process.

## Note
- Creating GitHub Release with Release Please is disabled (`skip-github-release: true`) to do releases with GoReleaser.
    - Therefore, there are several steps that have to be done manually, such as removing labels after release.
- I have not found a way to make a commit releasable, such as `chore`. Therefore, you have to manually trigger the creation of a release PR in order to create such a commit. See the documentation in the PR for more information.

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
@aquasecurity/trivy Do not hesitate to comment if anything is unclear, and suggest changes on the document.

**Comment by simar7** [comment]
Thanks for the writeup, I have a small question

> Merge releasable commits into the release branch (main and release/v*).

Will have two active development branches now? `main` and `release/v*`? Does that also mean we'll need duplicate each PR (to land on each release branch)? I think I'm most likely not correct in my assumption but if you can explain a little more, it would help.

**Comment by nikpivkin** [comment]
`Release Please` does not seem to support the dependabot message format. https://github.com/googleapis/release-please/issues/2295

Will a changelog be generated for all previous versions? Is there a need for this?

**Comment by knqyf263** [comment]
>Will have two active development branches now? main and release/v*? Does that also mean we'll need duplicate each PR (to land on each release branch)? I think I'm most likely not correct in my assumption but if you can explain a little more, it would help.

I am trying to keep it the same as the current cycle. Like today, we basically maintain the main branch. We'll create `release/v*` and backport patches to the branch only when we find critical bugs.

**Comment by knqyf263** [comment]
>Release Please does not seem to support the dependabot message format. https://github.com/googleapis/release-please/issues/2295

Yes, `chore(deps)` is not releasable in Release Please, and that's why I added a manual flow. However, this only means that a release PR is not created when there are only `chore:` commits, and a release PR is created when there are other `fix:` commits. Furthermore, when we only bump the version of dependent modules, we may not need to release a new version.

I think we should use `fix(deps): bump trivy-checks` when we explicitly want to fix a bug in trivy-checks or trivy-aws. What do you think?

>Will a changelog be generated for all previous versions? Is there a need for this?

I [set an initial version](https://github.com/aquasecurity/trivy/pull/6795/files#diff-e38bb01e41ed5baa3c9e8111b86009ca62d4498d2aefd14d367cd6300dd8b126R1). The changelog is created after the version.

File: .github/workflows/release-please.yaml
Comment (DmitriyLewen) [inline .github/workflows/release-please.yaml:12]: IIUC we don't need to use `v` prefix for version, because PR titile format contains it:
https://github.com/aquasecurity/trivy/blob/efc7946b2e03619c6464e65d63c204960553c13f/release-please-config.json#L7
Perhaps we need to write about this to avoid human error.

File: docs/community/maintainer/release.md
Comment (DmitriyLewen) [inline docs/community/maintainer/release.md:1]: `mkdocs.yaml` doesn't contain this page:
https://github.com/aquasecurity/trivy/blob/349caf96bc3dd81551d488044f1adfdb947f39fb/mkdocs.yml#L197-L199

File: docs/community/maintainer/release.md
Comment (DmitriyLewen) [inline docs/community/maintainer/release.md:29]: > Like today, we basically maintain the main branch. We'll create release/v* and backport patches to the branch only when we find critical bugs.

IIUC we will create `release/vX.Y` branch only for patches with bug fixes.

So we need to use the patch part in the branch name to add the ability to release more than one patch for a minor version.

```suggestion
The `release/vX.Y.Z` release branches are also subject to automatic release PR creation for patch releases.
```

File: docs/community/maintainer/release.md
Comment (knqyf263) [reply docs/community/maintainer/release.md:1]: Added https://github.com/aquasecurity/trivy/pull/6795/commits/fce855452e8d51dc18f0d816f1d9b2a8cca028c6

File: .github/workflows/release-please.yaml
Comment (knqyf263) [reply .github/workflows/release-please.yaml]: It's written in the doc, but it's a good idea also to add it here. Done. https://github.com/aquasecurity/trivy/pull/6795/commits/22abe912b02fd2376e114dd9313b8f18ed84efb6

File: docs/community/maintainer/release.md
Comment (DmitriyLewen) [reply docs/community/maintainer/release.md]: Oh... I think I've found a flaw in my logic

I wanted to split all the patches into separate branches.
But in this case, for patch `vX.Y.2` and above, we will skip changes from previous patches (or we will have to add them manually).

for `vX.Y` we will simply add the necessary commits and release patches when they are needed

Thanks for the clarification, now I understand this point 👍

**Review by DmitriyLewen — APPROVED** [review]

File: docs/community/maintainer/release.md
Comment (knqyf263) [reply docs/community/maintainer/release.md]: Creating a branch for a minor version is straightforward to me. We can backport patches from `main` to `release/vX.Y`, and cut vX.Y.1, vX.Y.2... on the branch. Many OSS projects are doing this, although their branch naming conventions are slightly different, `1.12`, `1.12.x`, `release-1.12`, `release-1.12.x`, `release/1.12`, etc.

**Comment by knqyf263** [comment]
@simar7 @chen-keinan @nikpivkin I'd wait for your feedback and approval.

After this change, anyone can release, so it is a good idea to appoint a person in charge of releases each month and have that person take the lead.

1. Ask maintainers to draft release notes first
2. Ask maintainers to review and approve the release PR
3. Merge the release PR
4. Write release notes when released

File: docs/community/maintainer/release-flow.md
Comment (chen-keinan) [inline docs/community/maintainer/release-flow.md:27]: @knqyf263 can you please elaborate on this part, does a release PR is created for every merged (feat / fix)? 
if so it will create a lot of release PR ?

Comment (knqyf263) [reply docs/community/maintainer/release-flow.md]: Added https://github.com/aquasecurity/trivy/pull/6795/commits/0e69b0f71ef96be4d7068579f2ea99b4925130f3. Thanks for your feedback!

File: docs/community/maintainer/release-flow.md
Comment (simar7) [inline docs/community/maintainer/release-flow.md:44]:
```suggestion
Specify the release version (without the `v` prefix) and click "Run workflow" to create a release PR for the specified version.
```

**Review by simar7 — APPROVED** [review]

**Pull request merged**
